File: kraken/src/main/java/glitch/kraken/services/request/StreamSearchRequest.java
package glitch.kraken.services.request;

import glitch.api.http.HttpClient;
import glitch.api.http.Routes;
import glitch.kraken.object.json.Stream;
import glitch.kraken.object.json.collections.Streams;
import glitch.service.AbstractRequest;
import reactor.core.publisher.Mono;

public class StreamSearchRequest extends AbstractRequest<Stream, Streams> {
    private Integer limit;
    private Integer offset;
    private Boolean hls;

    public StreamSearchRequest(HttpClient http, String query) {
        super(http, Routes.get("/search/channels").newRequest().queryParam("query", encodeQuery(query)));
    }

    public StreamSearchRequest setLimit(Integer limit) {
        this.limit = limit;
        return this;
    }

    public StreamSearchRequest setOffset(Integer offset) {
        this.offset = offset;
        return this;
    }

    public StreamSearchRequest setHls(Boolean hls) {
        this.hls = hls;
        return this;
    }

    @Override
    public Mono<Streams> get() {

        if (limit != null && limit > 0 && limit <= 100) {
            request.queryParam("limit", limit);
        }

        if (offset != null && offset >= 0) {
            request.queryParam("offset", offset);
        }

        if (hls != null) {
            request.queryParam("hls", hls);
        }

        return httpClient.exchangeAs(request, Streams.class);
    }
}

File: kraken/src/main/kotlin/io/glitchlib/v5/service/request/StreamSearchRequest.kt
package io.glitchlib.v5.service.request

import io.glitchlib.internal.http.AbstractRequest
import io.glitchlib.internal.http.HttpRequest

class StreamSearchRequest(private val query: String) : AbstractRequest {

    private var limit: Int? = null
    private var offset: Int? = null
    private var httpLiveStreaming: Boolean = false

    fun limit(limit: Int) = apply {
        this.limit = limit
    }

    fun offset(offset: Int) = apply {
        this.offset = offset
    }

    fun httpLiveStreaming(httpLiveStreaming: Boolean) = apply {
        this.httpLiveStreaming = httpLiveStreaming
    }

    override fun invoke(): HttpRequest.() -> Unit = {

        if (limit != null && limit!! > 0 && limit!! <= 100) {
            addQueryParameters("limit", limit!!.toString())
        }

        if (offset != null && offset!! >= 0) {
            addQueryParameters("offset", offset!!.toString())
        }

        if (httpLiveStreaming) {
            addQueryParameters("hls", httpLiveStreaming.toString())
        }
    }
}
